feat(client) ✨: Add min/max players check

SBAsai opened the pull request:
This PR will check if the minimum player count is met in order to start the current game session. It will also check if the maximum player count is met (lobby is full) and prevent more players from joining in. Each check will display an error accordingly.

SBAsai committed:
feat(client): ✨ Add minimum and maximum player count

Added a restriction to the minimum players required to start a game and a maximum players allowed in a lobby. Errors are thrown accordingly

SBAsai committed:
Merge branch 'development' into asai/lobby-minmax

SBAsai committed:
refactor(client): 🎨 Fix max player count and max player check

Changed the maximum number of player to 8 and changed the check to account for greater than the max players

SBAsai committed:
feat(client): ✨ Add min and max player count in lobby UI

Added the minimum players left text and maximum player count denominator to the lobby UI. Made max players and min players a variable in firestore types.

SBAsai committed:
Merge branch 'development' into asai/lobby-minmax

SBAsai committed:
refactor(client): ♻️ Change minimum player text class

Changed the minimum player text class to now be `error`

nstringham commented on functions/src/firestore-types/lobby.ts:
```diff
@@ -92,6 +92,16 @@ export type Lobby = {
    * @note this can be a uid, NONE, or undefined
    */
   votedOff?: string | "NONE";
+
+  /**
+   * the maximum players allowed in a current lobby session
+   */
+  maxPlayers: number;
+
+  /**
+   * the minimum players required to start a game session
+   */
+  minPlayers: number;
```
why are these not constants? I don't think these are configurable

SBAsai replied:
You mean like hard-coded values? I figured we may want to change the numbers depending on the number of catfish

SBAsai replied:
I updated this comment section to describe the reason it needs to be configurable for now.

nstringham commented on functions/src/lobby.ts:
```diff
@@ -87,11 +89,16 @@ export const startGame = functions.https.onCall(async (data: unknown, context):
 
+    // throw an error if there aren't enough players in the lobby
+    if (uids.length < minPlayers) {
+      throw new functions.https.HttpsError("permission-denied", "Not enough players to start the game!");
```
this should probobly be `failed-precondition`

nstringham commented on functions/src/lobby.ts:
```diff
@@ -125,11 +132,16 @@ export const joinLobby = functions.https.onCall((data: unknown, context): Promis
 
+    // throw an error if the lobby is already full
+    if (uids.length >= maxPlayers) {
+      throw new functions.https.HttpsError("permission-denied", "Lobby is full!");
```
this should probobly be `failed-precondition`

nstringham commented on client/src/components/Lobby.svelte:
```diff
@@ -9,6 +9,7 @@
   import { changeAvatar, startGame, leaveLobby } from "$lib/firebase/firebase-functions";
   import { goto } from "$app/navigation";
   import { auth } from "$lib/firebase/app";
+  import { missing_component } from "svelte/internal";
```
I don't think you want to do this

SBAsai replied:
It's not being used so I think VSCode just auto imported it when I was doing something

SBAsai committed:
refactor(functions): ♻️ Change HTTPS error code

Changed HTTPS error code from permission-denied to failed-precondition

SBAsai committed:
refactor(client): ♻️ Remove unused import

Removed unused import from `Lobby.svelte`

SBAsai committed:
refactor(functions): ♻️ Update variable comments

Updated the variable description for min and max players to better explain the values

SBAsai committed:
Merge branch 'development' into asai/lobby-minmax

SBAsai committed:
refactor(client): ♻️ Add placeholder firestore type

Changed min and max players from being a firestore type and replaced them with `catfishAmount` which will appear with the lobby settings component

SBAsai committed:
style(client): 🎨 Fix prettier warning

Formatted with prettier

nstringham commented on client/src/components/Lobby.svelte:
```diff
@@ -69,7 +70,18 @@
+        <!-- Grammar check -->
+        <h3 class="error">{minPlayers - lobby.players.length} more players required to start game...</h3>
+      {:else}
+        <h3 class="error">{minPlayers - lobby.players.length} more player required to start game...</h3>
```
this could be `1` instead of `{minPlayers - lobby.players.length}`

nstringham commented on client/src/components/Lobby.svelte:
```diff
@@ -14,6 +14,7 @@
   export let lobbyCode: string;
   export let lobby: Lobby;
   let errorMessage: string = "";
+  let minPlayers = lobby.catfishAmount * 2 + 2;
```
this should be reactive incase the host changes the game setting

nstringham reviewed (dismissed):
the start game button should be disabled if the game cannot be started

SBAsai committed:
refactor(client): 🎨 Change client side variable calls

Changed `minPlayers` from being called to 1 when there is one player needed left. Made `minPlayers` reactive

SBAsai committed:
style(client): 🎨 Add button disable on start game button

Made the start game button disabled if the minium players requirement is not met

SBAsai committed:
Merge branch 'development' into asai/lobby-minmax

gchamb commented:
`on line 2 ` in Lobby.svelte you have an unused component

gchamb commented on client/src/components/Lobby.svelte:
```diff
@@ -69,12 +70,23 @@
+    {/if}
   </div>
   <SelectAvatar {lobby} on:change={(event) => onAvatarSelect(event.detail.value)} />
   {#if auth.currentUser?.uid === lobby.uids[0]}
```
This probably wasn't you but we should be using the authstore here

gchamb replied:
`on line 11` use the authstore. I know that probably wasn't you who added it

gchamb reviewed (dismissed):
see above

SBAsai committed:
refactor(client): ♻️ Remove unused component

Removed unused component

SBAsai committed:
refactor(client): ♻️ Use authstore

Changed `auth.user` call to `$user` using authstore. Formatted with prettier
